### PR TITLE
add SpawnEntities and followFloor

### PR DESCRIPTION
**Pull request by tetorea**
- add the SpawnEntities ros service
- add the ability for RobotVehicleMovementComponent to (1) move above the floor based on contact points, and (2) be moved by moving platforms  

To test with [Obayashi_UE PR](https://github.com/rapyuta-robotics/obayashi_UE/pull/16)

### CODE REVIEW

**Review by yuokamoto — APPROVED** [review]

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 1]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
-    FQuat DeltaRotation(FVector::ZAxisVector, InversionFactor * AngularVelocity.Z * InDeltaTime);
+    FQuat deltaRotation(FVector::ZAxisVector, InversionFactor * AngularVelocity.Z * InDeltaTime);
+
+    DesiredRotation = oldRotation * deltaRotation;
+    DesiredMovement = (oldRotation * position);
+
+    // if Robot is on a moving platform, add the platform motion
+    if (MovingPlatform != nullptr && bFollowPlatform)
+    {
+        FVector CurrentPlatformLocation = FVector(MovingPlatform->GetActorLocation());
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp:32]:
```suggestion
        FVector currentPlatformLocation = MovingPlatform->GetActorLocation();
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 2]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+        FVector PlatformTranslation = CurrentPlatformLocation - LastPlatformLocation;
+        DesiredMovement += PlatformTranslation;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
        FVector currentPlatformLocation = MovingPlatform->GetActorLocation();
        FVector platformTranslation = currentPlatformLocation - LastPlatformLocation;
        DesiredMovement += platformTranslation;

```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 3]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+            PawnOwner->SetActorRotation(UKismetMathLibrary::MakeRotFromXZ(ForwardProjection, PlaneNormal));
+
+            if (MovingPlatform == nullptr)
+            {
+                // Moves the robot up or down, depending on impact position
+                float MinDistance = ContactsDistance[0];
+                if (ContactsDistance[1] < MinDistance)
+                    MinDistance = ContactsDistance[1];
+                if (ContactsDistance[2] < MinDistance)
+                    MinDistance = ContactsDistance[2];
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
                float minDistance = Algo::MinElement(ContactsDistance);
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 4]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+
+                MinDistance -= RayOffsetUp;
+
+                if (MinDistance > FallingSpeed * InDeltaTime)
+                    MinDistance = FallingSpeed * InDeltaTime;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
                 MinDistance = FMath::Min(MinDistance, FallingSpeed * InDeltaTime);
```
so the multiplication is done once only.

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 5]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+
+                FVector HeightVariation = {0., 0., -MinDistance};
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
                FVector heightVariation = {0.f, 0.f, -MinDistance};
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 6]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+                        ContactsDistance[MaxContactDistanceIndex] = Hit.Distance;
+                        Contacts[MaxContactDistanceIndex] = Hit.ImpactPoint;
+                    }
+                }
+            }
+
+            // get the normal vector of the plane going through these 3 points
+            FVector PlaneNormal = FVector::CrossProduct(Contacts[1] - Contacts[0], Contacts[2] - Contacts[0]);
+            PlaneNormal.Normalize(0.01);
+            if (PlaneNormal.Z < 0.)
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
            PlaneNormal.Normalize(0.01f);
            if (PlaneNormal.Z < 0.f)
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 7]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+                bool IsFloorHit = GetWorld()->LineTraceSingleByChannel(Hit,
+                                                                       startPos,
+                                                                       endPos,
+                                                                       ECollisionChannel::ECC_Visibility,
+                                                                       TraceParams,
+                                                                       FCollisionResponseParams::DefaultResponseParam);
+                if (!IsFloorHit)
+                {
+                    // if no impact below, just consider this contact point is falling
+                    Hit.ImpactPoint = Contact->GetComponentLocation() - FVector(0., 0., FallingSpeed * InDeltaTime);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
                    Hit.ImpactPoint = Contact->GetComponentLocation() - FVector(0.f, 0.f, FallingSpeed * InDeltaTime);
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 8]
```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
...
+            // get the normal vector of the plane formed by these 3 points
+            FVector Contacts[3];
+            float ContactsDistance[3];
+            uint8 NbContact = 0;
+
+            for (USceneComponent* Contact : ContactPoints)
+            {
+                // get the contact points
+                FVector startPos = Contact->GetComponentLocation() + FVector(0., 0., RayOffsetUp);
+                FVector endPos = Contact->GetComponentLocation() - FVector(0., 0., RayOffsetDown);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
                FVector startPos = Contact->GetComponentLocation() + FVector(0.f, 0.f, RayOffsetUp);
                FVector endPos = Contact->GetComponentLocation() - FVector(0.f, 0.f, RayOffsetDown);
```

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 9]
```diff
@@ -152,3 +289,64 @@ FTransform URobotVehicleMovementComponent::GetOdomTF() const
...
+
+    FVector startPos = PawnOwner->GetActorLocation() + FVector(0., 0., 10.);
+    FVector endPos = PawnOwner->GetActorLocation() - FVector(0., 0., 50.);
+    FHitResult HitResult;
+    bool IsFloorHit = GetWorld()->LineTraceSingleByChannel(HitResult,
+                                                           startPos,
+                                                           endPos,
+                                                           ECollisionChannel::ECC_Visibility,
+                                                           TraceParams,
+                                                           FCollisionResponseParams::DefaultResponseParam);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]: maybe not clang-formatted yet?

File: Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp [hunk 10]
```diff
@@ -152,3 +289,64 @@ FTransform URobotVehicleMovementComponent::GetOdomTF() const
...
+               MinDistanceToFloor);
+    }
+}
+
+void URobotVehicleMovementComponent::SetMovingPlatform(AActor* InPlatform)
+{
+    // UE_LOG(LogTemp, Warning, TEXT("URobotVehicleMovementComponent::SetMovingPlatform..."));
+    MovingPlatform = InPlatform;
+    LastPlatformLocation = FVector(InPlatform->GetActorLocation());
+    LastPlatformRotation = FQuat(InPlatform->GetActorQuat());
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp:340]:
```suggestion
    LastPlatformLocation = InPlatform->GetActorLocation();
    LastPlatformRotation = InPlatform->GetActorQuat();

```

Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp]:
```suggestion
    LastPlatformLocation = InPlatform->GetActorLocation();
    LastPlatformRotation = InPlatform->GetActorQuat();
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h [hunk 11]
```diff
@@ -28,15 +29,26 @@ class RCLUE_API URobotVehicleMovementComponent : public UPawnMovementComponent
...
+    UPROPERTY(VisibleAnywhere)
+    AActor* MovingPlatform = nullptr;    // The platform below the robot
 
-    UPROPERTY(Transient)
-    FQuat DesiredRotation = FQuat::Identity;
+    UPROPERTY(VisibleAnywhere)
+    FVector LastPlatformLocation;
+
+    UPROPERTY(VisibleAnywhere)
+    FQuat LastPlatformRotation;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h]:
```suggestion
    UPROPERTY(VisibleAnywhere)
    FVector LastPlatformLocation = FVector::ZeroVector;

    UPROPERTY(VisibleAnywhere)
    FQuat LastPlatformRotation = FQuat::Identity;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h [hunk 12]
```diff
@@ -45,13 +57,24 @@ class RCLUE_API URobotVehicleMovementComponent : public UPawnMovementComponent
...
+    FString ChildFrameId = TEXT("base_footprint");
 
     void SetFrameIds(const FString& InFrameId, const FString& InChildFrameId);
 
     UPROPERTY(EditAnywhere)
     FTransform InitialTransform = FTransform::Identity;
 
+    // For slopes, complex floors, free fall
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float RayOffsetUp = 10.;    // Ray start Z offset. Value must be > possible penetration of objects in contact point, in one tick
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h:69]:
```suggestion
    float RayOffsetUp = 10.f;    // Ray start Z offset. Value must be > possible penetration of objects in contact point, in one tick

```

File: Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h [hunk 13]
```diff
@@ -45,13 +57,24 @@ class RCLUE_API URobotVehicleMovementComponent : public UPawnMovementComponent
...
+
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float RayOffsetDown = 20.;    // Ray end Z offset
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h:72]:
```suggestion
    float RayOffsetDown = 20.f;    // Ray end Z offset
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h [hunk 14]
```diff
@@ -65,7 +88,27 @@ class RCLUE_API URobotVehicleMovementComponent : public UPawnMovementComponent
...
+    UFUNCTION(BlueprintCallable)
+    void RemoveMovingPlatform();
+
+    // For slopes, complex floors, free fall
+    UPROPERTY(VisibleAnywhere)
+    float MinDistanceToFloor =
+        0.f;    // Z distance between the robot root location and the floor, used when less than 3 contact points are defined
+
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float FallingSpeed = 100.;    // How much the robot falls if no floor beneath ( FallingSpeed * DeltaTime )
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h]:
```suggestion
    float FallingSpeed = 100.f;    // How much the robot falls if no floor beneath ( FallingSpeed * DeltaTime )

```

**Review by Tadinu — APPROVED** [review]
Thanks for the substantial PR with useful updates, just minor comments about general coding conventions:
* Local var in camelCase (as opposed to member in PascalCase)
* bool var/member prefixed with `b`
* float constant postfized with `.f` for not being store as double
* Member var should be initialized just for clarity due to UE maybe not auto-doing it

File: Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h [hunk 15]
```diff
@@ -45,13 +57,24 @@ class RCLUE_API URobotVehicleMovementComponent : public UPawnMovementComponent
...
     UPROPERTY(EditAnywhere)
     FTransform InitialTransform = FTransform::Identity;
 
+    // For slopes, complex floors, free fall
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float RayOffsetUp = 10.;    // Ray start Z offset. Value must be > possible penetration of objects in contact point, in one tick
+
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float RayOffsetDown = 20.;    // Ray end Z offset
+    // Rays go from ContactPoint+RayOffsetUp to ContactPoint-RayOffsetDown
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RobotVehicleMovementComponent.h]:
```suggestion
    UPROPERTY(EditAnywhere, BlueprintReadWrite)
    float RayOffsetUp = 10.f;    // Ray start Z offset. Value must be > possible penetration of objects in contact point, in one tick

    UPROPERTY(EditAnywhere, BlueprintReadWrite)
    float RayOffsetDown = 20.f;    // Ray end Z offset
    // Rays go from ContactPoint+RayOffsetUp to ContactPoint-RayOffsetDown
```

**Pull request merged**